src/Eloquent/Builder.php

```diff
@@ -177,22 +180,27 @@ public function raw($value = null)
         $results = $this->query->raw($value);
 
         // Convert MongoCursor results to a collection of models.
-        if ($results instanceof Cursor) {
-            $results = iterator_to_array($results, false);
+        if ($results instanceof CursorInterface) {
+            $results->setTypeMap(['root' => 'array', 'document' => 'array', 'array' => 'array']);
+            $results = $this->query->aliasIdForResult(iterator_to_array($results));
 
             return $this->model->hydrate($results);
         }
 
-        // Convert MongoDB BSONDocument to a single object.
-        if ($results instanceof BSONDocument) {
-            $results = $results->getArrayCopy();
-
-            return $this->model->newFromBuilder((array) $results);
+        // Convert MongoDB Document to a single object.
+        if (is_object($results) && (property_exists($results, '_id') || property_exists($results, 'id'))) {
```

Comment:
Confirmed that PHP correctly calls the `has_property` handler when using `property_exists`: https://github.com/php/php-src/blob/1be989bbf0e46db8438384db184fca424392cfef/Zend/zend_builtin_functions.c#L1020..L1023